test/Base_Tests/src/Data/Round_Spec.enso

```diff
@@ -6,276 +6,328 @@ from Standard.Test import all
 
 polyglot java import java.math.BigInteger
 
+type Batch_Runner
+    Value (run_batch : Vector -> Vector -> Boolean -> Vector) (run_one : Number -> Integer -> Boolean -> Number)
+
+    from_function (f : Number -> Integer -> Boolean -> Number) =
+        batch values_vec dps_vec use_bankers =
+            values_vec.zip dps_vec v-> dp-> f v dp use_bankers
+        Batch_Runner.Value batch f
+
+    run self (use_bankers : Boolean) (action : Batch_Builder -> Nothing) =
```

Comment:
```suggestion
    run self (use_bankers : Boolean) (action : (Number -> Integer -> Check_Instance) -> Nothing) =
```

What is the advantage of batching the operations like this? How does it make it more efficient?

Reply:
In Snowflake that is running in the cloud, the minimum latency from my computer to the Frankfurt data center is ~15-20ms. Usually the Snowflake query overhead totals about 70ms. Merely executing the computation is much less - that's why e.g. Postgres tests were much faster.

With batching we pay the roundtrip/overall query overhead only once per batch, so the execution time is faster.

Reply:
You made me check the efficiency and ooops - there's basically no significant difference!

| Test | Without batching | With batching |
|----------------------------------------|-------:|--:|
| Can round positive decimals correctly  | 1409ms | 1107ms |
| Can round negative decimals correctly  | 1013ms | 2261ms |
| Explicit and implicit 0 decimal places work the same | 606ms | 395ms |
| Can round zero and small decimals correctly | 509ms | 1233ms |
| Can round positive decimals to a specified number of decimal places | 2011ms | 3930ms |
| Can round negative decimals to a specified number of decimal places | 1808ms | 3504ms |
| Can round positive decimals to a specified negative number of decimal places | 1656ms | 1368ms |
| Can round negative decimals to a specified negative number of decimal places | 1743ms | 949ms |

So the problem is - when I call `round` on a `DB_Column`, it checks its `value_type`, triggering a small no-results query that asks the DB for the type.

Thus regardless if I batch run N rounds on separate queries or batch them in a single query, I still get: N+1 queries with batching (N type checks and 1 more complex execute) vs 2*N queries without batching (type check + execute for each operation). Apparently, timing wise the result is similar.

I want to fix the batching and see if it gets better.

Reply:
So in [f9bbbe0](https://github.com/enso-org/enso/pull/10458/commits/f9bbbe08110cf9ab211dcd3dca57861c9c7f2a4f) I have changed how we work with the `SQL_Type_Reference` - now if a statement containing a column with that type is read in, its type is cached into the reference - so that a separate query for reading the reference is avoided. We only perform a separate types-only query if the type is checked before querying.

With that fix the results are much better:
| Test | Without batching (old measurement) | With fixed batching |
|----------------------------------------|-------:|--:|
| Can round positive decimals correctly  | 1409ms | 673ms |
| Can round negative decimals correctly  | 1013ms | 253ms |
| Explicit and implicit 0 decimal places work the same | 606ms | 207ms |
| Can round zero and small decimals correctly | 509ms | 165ms |
| Can round positive decimals to a specified number of decimal places | 2011ms | 291ms |
| Can round negative decimals to a specified number of decimal places | 1808ms | 371ms |
| Can round positive decimals to a specified negative number of decimal places | 1656ms | 419ms |
| Can round negative decimals to a specified negative number of decimal places | 1743ms | 423ms |

Now the speed-up is 2-6x so the batching is worth it. With also disabling of some edge case tests, the overall Rounding part is down from ~60s to 4s.